### ISSUE TITLE
Database creation died with pandas error

### ISSUE THREAD

**Issue by jeankeller** [qiyunlab/HGTector#92]
Hi,

Very interesting tool, I tried to use it with the following command : 
hgtector database --output dbTest/ --cats bacteria,archaea,fungi,protozoa,viral,plant -t 1117,766 -e --sample 1 --rank species_latin --above --reference --represent --typemater --compile diamond
But after downloading the data, the script crashed with the following error:

> `Traceback (most recent call last):
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/scope.py", line 193, in resolve
>     return self.scope[key]
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/collections/__init__.py", line 986, in __getitem__
>     return self.__missing__(key)            # support subclasses that define __missing__
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/collections/__init__.py", line 978, in __missing__
>     raise KeyError(key)
> KeyError: 'failures'
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/scope.py", line 208, in resolve
>     return self.temps[key]
> KeyError: 'failures'
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/home/jeankeller/anaconda3/envs/hgtector2/bin/hgtector", line 96, in <module>
>     main()
>   File "/home/jeankeller/anaconda3/envs/hgtector2/bin/hgtector", line 35, in main
>     module(args)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/hgtector/database.py", line 151, in __call__
>     self.download_genomes()
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/hgtector/database.py", line 638, in download_genomes
>     self.df.query('genome in @failures', inplace=True)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/frame.py", line 4111, in query
>     res = self.eval(expr, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/frame.py", line 4240, in eval
>     return _eval(expr, inplace=inplace, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/eval.py", line 350, in eval
>     parsed_expr = Expr(expr, engine=engine, parser=parser, env=env)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 811, in __init__
>     self.terms = self.parse()
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 830, in parse
>     return self._visitor.visit(self.expr)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 415, in visit
>     return visitor(node, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 421, in visit_Module
>     return self.visit(expr, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 415, in visit
>     return visitor(node, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 424, in visit_Expr
>     return self.visit(node.value, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 415, in visit
>     return visitor(node, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 723, in visit_Compare
>     return self.visit(binop)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 415, in visit
>     return visitor(node, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 536, in visit_BinOp
>     op, op_class, left, right = self._maybe_transform_eq_ne(node)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 458, in _maybe_transform_eq_ne
>     right = self.visit(node.right, side="right")
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 415, in visit
>     return visitor(node, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/expr.py", line 549, in visit_Name
>     return self.term_type(node.id, self.env, **kwargs)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/ops.py", line 98, in __init__
>     self._value = self._resolve_name()
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/ops.py", line 115, in _resolve_name
>     res = self.env.resolve(self.local_name, is_local=self.is_local)
>   File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/pandas/core/computation/scope.py", line 213, in resolve
>     raise UndefinedVariableError(key, is_local) from err
> pandas.core.computation.ops.UndefinedVariableError: local variable 'failures' is not defined`
> 

Could it be due to the Pandas (1.4.3)/Python (3.10) version from the Conda virtual env? Virtual env has been created as described in the installation section.

Thank you,

Jean

**Comment by qiyunzhu** [comment]
@jeankeller Thanks for reporting this issue! I checked the code and found that it was a bug. I just fixed it. Now you can update the program and run it again to see if the problem has been fixed. Thank you!

To reinstall HGTector:

```bash
pip install --force-reinstall --no-cache-dir git+https://github.com/qiyunlab/HGTector.git
```

**Comment by jeankeller** [comment]
Thanks for the new version, the problem is fixed but a new error appeared. 
When HGTector tries to create database, some genomes are not found provoking the following error : 

> Traceback (most recent call last):
  File "/home/jeankeller/anaconda3/envs/hgtector2/bin/hgtector", line 96, in <module>
    main()
  File "/home/jeankeller/anaconda3/envs/hgtector2/bin/hgtector", line 35, in main
    module(args)
  File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/hgtector/database.py", line 154, in __call__
    self.extract_genomes()
  File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/site-packages/hgtector/database.py", line 671, in extract_genomes
    with gzip.open(lfile, 'rb') as f:
  File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/gzip.py", line 58, in open
    binary_file = GzipFile(filename, gz_mode, compresslevel)
  File "/home/jeankeller/anaconda3/envs/hgtector2/lib/python3.10/gzip.py", line 174, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'db0507/download/faa/GCF_000883615.1_ViralProj34737_protein.faa.gz'

I had a look to few of these unfound genomes and it seems these are species with extremely reduced genomes without predicted proteins. Could it be the source of the problem?

**Comment by qiyunzhu** [comment]
Hello @jeankeller Sorry for the late response. Had other work responsibilities. Thanks for reporting the bug!

Many viral genomes don't have pre-annotated proteins (like the [one](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/883/615/GCF_000883615.1_ViralProj34737/) you reported). I overlooked this problem earlier. Now the code has been fixed and should work. You may reinstall the program as stated above. To quickly test whether the patch works, you can do:

```bash
hgtector database -c viral -o . --compile none -s 1 -r family
```
You will see that a few viral genomes don't have proteins. But the program correctly ignores them and takes care of the remaining ones.

Then you can re-run the entire workflow. Hope this helps!

**Comment by jeankeller** [comment]
Hello @qiyunzhu, thank you for the fix, it works now. But I had to manually compiled the database because an error occured when the --compile option was set (manual compilation with the same Diamond executable as the one provided to HGTECTOR).
Thank you again for your support

**Comment by qiyunzhu** [comment]
Hi @jeankeller Thanks for the update. I now fixed the `--compile diamond` issue in #97 . The automatic database compilation should work. I am glad that you have already solved the problem by manual compilation. But in case you want to try, you can update the program and try it.
